sycl/test-e2e/CompositeDevice/composite_device.cpp

```diff
@@ -0,0 +1,228 @@
+
+using namespace sycl::ext::oneapi::experimental;
+
+bool isL0Backend(sycl::backend backend) {
+  return (backend == sycl::backend::ext_oneapi_level_zero);
+}
+
+bool isCombinedMode() {
+  const char *Mode = std::getenv("ZE_FLAT_DEVICE_HIERARCHY");
+  return (Mode != nullptr) && (std::strcmp(Mode, "COMBINED") == 0);
+}
+
+int main() {
+  sycl::queue q;
+  bool IsCombined = isCombinedMode();
+  auto Platforms = sycl::platform::get_platforms();
+
+  // Check that device::get_devices() and platform::get_devices() do not return
+  // composite devices.
+  {
+    auto Devs = sycl::device::get_devices();
+    for (const auto &D : Devs) {
+      assert(!D.has(sycl::aspect::ext_oneapi_is_composite));
+
+      // If ZE_FLAT_DEVICE_HIERARCHY != COMBINED,
+      // sycl::aspect::ext_oneapi_is_component must be false.
+      assert(!(!IsCombined && D.has(sycl::aspect::ext_oneapi_is_component)));
+    }
+
+    for (const auto &P : Platforms) {
+      bool IsL0 = isL0Backend(P.get_backend());
+      if (!IsL0)
+        continue;
+
+      Devs = P.get_devices();
+      for (const auto &D : Devs) {
+        assert(!D.has(sycl::aspect::ext_oneapi_is_composite));
+
+        // If ZE_FLAT_DEVICE_HIERARCHY != COMBINED,
+        // sycl::aspect::ext_oneapi_is_component must be false.
+        assert(!(!IsCombined && D.has(sycl::aspect::ext_oneapi_is_component)));
+      }
+    }
+  }
+
+  // Check that:
+  //   A. The free function get_composite_devices returns all of the composite
+  //      devices across all platforms.
+  //   B. The member function platform::ext_oneapi_get_composite_devices returns
+  //      the composite devices within the given platform.
+  //   C. The APIs defined in this extension are only useful when using the
+  //      Level Zero backend, and they are only useful when the Level Zero
+  //      environment variable ZE_FLAT_DEVICE_HIERARCHY=COMBINED is set. The
+  //      APIs may be called even when using other backends, but they will
+  //      return an empty list of composite devices.
+  //   D. The execution environment for a SYCL application has a fixed number of
+  //      composite devices which does not vary as the application executes. As
+  //      a result, each call to these functions returns the same set of device
+  //      objects, and the order of those objects does not vary between calls.
+  {
+    std::vector<sycl::device> AllCompositeDevs = get_composite_devices();
+    std::vector<sycl::device> CombinedCompositeDevs;
+    for (const auto &P : Platforms) {
+      auto CompositeDevs = P.ext_oneapi_get_composite_devices();
+      bool IsL0 = isL0Backend(P.get_backend());
+      // Check C.
+      if (!IsL0 || !IsCombined)
+        assert(CompositeDevs.empty());
+
+      for (const auto &D : CompositeDevs) {
+        if (std::find(CombinedCompositeDevs.begin(),
+                      CombinedCompositeDevs.end(),
+                      D) == CombinedCompositeDevs.end())
+          CombinedCompositeDevs.push_back(D);
```

Comment:
Why do we need this check? Devices across platforms can't be the same and I would not expect to have duplicate devices in `P.ext_oneapi_get_composite_devices()`.

Reply:
I'm just checking that the following statement from extension spec holds for our implementation:

> The free function `get_composite_devices` returns all of the composite devices across all platforms. The member function `platform::ext_oneapi_get_composite_devices` returns the composite devices within the given platform.

Particularly, we are checking that the free function returns the composite devices across all platforms, and we are not missing any due to a bug.